pkg/ebpf/map.go

```diff
@@ -70,13 +71,42 @@ func (m *Map) OpenOrCreate() error {
 	mapType := bpf.GetMapType(bpf.MapType(m.spec.Type))
 	m.spec.Flags = m.spec.Flags | bpf.GetPreAllocateMapFlags(mapType)
 
+	path := bpf.MapPath(m.spec.Name)
+
 	newMap, err := ciliumebpf.NewMapWithOptions(m.spec, opts)
 	if err != nil {
-		return fmt.Errorf("unable to create map: %w", err)
+		if !errors.Is(err, ciliumebpf.ErrMapIncompatible) {
+			return fmt.Errorf("unable to create map: %w", err)
+		}
+
+		// There already exists a pinned map but it has a different
+		// configuration (e.g different type, k/v size or flags).
+		// Try to delete and recreate it.
```

Comment:
Is it safe to do this for all maps? I also assume this can't be done "atomically"?

Reply:
> Is it safe to do this for all maps?

That's what we are already doing with all the other maps that are opened using the `bpf.OpenOrCreate()` method:

https://github.com/cilium/cilium/blob/648f4f05c46abb98d3f31bd5610db283d0cd3953/pkg/bpf/map_linux.go#L503-L505

I'm not saying that this is necessarily the best approach, but since the aim of the `ebpf` package is to (at least initially) mimic the `bpf` package (to avoid introducing another dimension of complexity while we migrate to that) I think we should stick with this approach for now

> I also assume this can't be done "atomically"?

That's my understanding, yes